### DISCUSSION TITLE
Using Streaming Algorithms that Require matrix_real

### DISCUSSION THREAD

**Discussion by nvssynthesis** [MTG/essentia#1389] [Q&A]
I am wondering how to use any of the streaming algorithms that require a matrix_real as one of their inputs or outputs. 
For example, the Onsets algorithm expects a matrix_real, which is a typedef for TNT::Array2D<float>, as the "detections" input. 
Normally, if it were e.g. a Vector<Real> input, I could use essentia::streaming::VectorInput<Real> to get it working in a streaming Network, but I am at a loss as how to do something similar with a matrix_real.

**Comment by Hikari-Tsai** [comment]
The streaming `Onsets` class is a `StreamingAlgorithmWrapper` around the standard algorithm. It does not build the detection matrix incrementally: it consumes **one matrix token** plus one weights-vector token, runs the standard algorithm, and emits one vector of onset times.

To feed it from C++, wrap the matrix in a vector so that `VectorInput` emits the whole matrix as one token:

```cpp
using Matrix = TNT::Array2D<essentia::Real>;

Matrix detections(numberFunctions, numberFrames);
std::vector<Matrix> detectionTokens{detections};
std::vector<std::vector<essentia::Real>> weightTokens{{1.0, 1.0}};

essentia::streaming::VectorInput<Matrix> matrixInput(&detectionTokens);
essentia::streaming::VectorInput<std::vector<essentia::Real>> weightsInput(
    &weightTokens
);

matrixInput.output("data") >> onsets->input("detections");
weightsInput.output("data") >> onsets->input("weights");
```

Do not use `VectorInput<Real>` for this port: that emits scalar tokens. Likewise, `VectorInput<std::vector<Real>>` is appropriate for a stream of vector tokens, but `detections` specifically requires a `TNT::Array2D<Real>` token.

If your upstream network naturally produces one detection vector per frame, collect those vectors first, transpose them into `[detection_function][frame]`, then send the completed matrix. `Onsets` itself is batch/global despite being exposed in the streaming namespace.
